parity-clib/parity.h

```diff
@@ -14,166 +14,270 @@
+    defined parity_logger || defined parity_config
+#error "macro conflicts with Parity Ethereum C API header"
+#endif
+
+typedef void (*parity_destructor)(void *);
 
 /// Parameters to pass to `parity_start`.
 struct ParityParams {
-	/// Configuration object, as handled by the `parity_config_*` functions.
-	/// Note that calling `parity_start` will destroy the configuration object (even on failure).
-	void *configuration;
-
-	/// Callback function to call when the client receives an RPC request to change its chain spec.
-	///
-	/// Will only be called if you enable the `--can-restart` flag.
-	///
-	/// The first parameter of the callback is the value of `on_client_restart_cb_custom`.
-	/// The second and third parameters of the callback are the string pointer and length.
-	void (*on_client_restart_cb)(void* custom, const char* new_chain, size_t new_chain_len);
-
-	/// Custom parameter passed to the `on_client_restart_cb` callback as first parameter.
-	void *on_client_restart_cb_custom;
-
-	/// Logger object which must be created by the `parity_config_logger` function
-	void *logger;
-};
+  /// Configuration object, as handled by the `parity_config_*` functions.
+  /// Note that calling `parity_start` will destroy the configuration object
+  /// (even on failure).
+  struct parity_config *configuration;
 
-#ifdef __cplusplus
-extern "C" {
-#endif
+  /// Callback function to call when the client receives an RPC request to
+  /// change its chain spec.
+  ///
+  /// Will only be called if you enable the `--can-restart` flag.
+  ///
+  /// The first parameter of the callback is the value of
+  /// `on_client_restart_cb_custom`. The second and third parameters of the
+  /// callback are the string pointer and length.
+  void (*on_client_restart_cb)(void *custom, const char *new_chain,
+                               uintptr_t new_chain_len);
+
+  /// Custom parameter passed to the `on_client_restart_cb` callback as first
+  /// parameter.
+  void *on_client_restart_cb_custom;
+
+  /// Callback for when a client is destroyed.
+  /// If this is NULL, no callback is called.
+  parity_destructor on_client_destroy;
+
+  /// Logger object which must be created by the `parity_config_logger` function
+  struct parity_logger *logger;
+};
 
 /// Builds a new configuration object by parsing a list of CLI arguments.
 ///
-/// The first two parameters are string pointers and string lengths. They must have a length equal
-/// to `len`. The strings don't need to be zero-terminated.
+/// The first two parameters are string pointers and string lengths. They must
+/// have a length equal to `len`. The strings don't need to be zero-terminated.
 ///
-/// On success, the produced object will be written to the `void*` pointed by `out`.
+/// On success, the produced object will be written to the `struct parity_config
+/// *` pointed by `out`.
 ///
-/// Returns 0 on success, and non-zero on error.
+/// Returns false on success, and true on error.
```

Comment:
Idiomatic C usually involves returning integers, so that you can pass back error codes.
I didn't return error codes because that was too cumbersome to define, but ideally that's IMO what we should do.

Reply:
I didn’t think of that.

Reply:
Fixed.